lib/Partitioner/PartitionerUtils.cpp

```diff
@@ -140,72 +139,154 @@ getOutUsersWithOnePredecessor(const std::set<Node *> &nodes) {
+  GraphMemInfo ret = info;
+
+  // Collect the Storage nodes used in this nodes set.
+  std::set<Storage *> usedStorage;
+  for (auto N : currNodes) {
+    for (size_t i = 0, e = N->getNumInputs(); i < e; i++) {
+      Storage *in = llvm::dyn_cast<Storage>(N->getNthInput(i).getNode());
+      if (in) {
+        usedStorage.insert(in);
+      }
+    }
+  }
+
+  std::set<NodeValue> usedNodeValue;
+  // The memory usage changes due to the node's inputs:
+  for (size_t i = 0, e = node->getNumInputs(); i < e; i++) {
+    if (llvm::isa<SaveNode>(node) && i == SaveNode::OutputIdx) {
       continue;
     }
-    // Check the inputs of each node in this subgraph and decide if it
-    // contributes to the memory usage:
-    for (size_t i = 0, e = cur->getNumInputs(); i < e; i++) {
-      NodeValue nodeVal = cur->getNthInput(i);
-      Node *node = nodeVal.getNode();
-      if (nodes.count(node) || nSet.count(nodeVal)) {
-        // This input belongs to this subgraph or it has been considered
-        // already, nothing to do.
+    NodeValue nodeVal = node->getNthInput(i);
+    Node *N = nodeVal.getNode();
+
+    if (usedNodeValue.count(nodeVal)) {
+      // This input has been considered already, nothing to do.
+      continue;
+    }
+
+    Storage *in = llvm::dyn_cast<Storage>(N);
+    if (in) {
+      if (usedStorage.count(in) != 0) {
         continue;
       }
-      nSet.insert(nodeVal);
-      Storage *in = llvm::dyn_cast<Storage>(node);
-      if (in) {
-        uint64_t size = in->getType()->getSizeInBytes();
-        if (node->getKind() == Kinded::Kind::ConstantKind) {
-          // Constant.
-          ret.constMemSize += size;
-        } else {
-          // PlaceHolder for Input.
-          ret.inMemSize += size;
-        }
+      // Node uses placeholders or constants which are not used in this set
+      // before, need to add the memory.
+      uint64_t size = in->getType()->getSizeInBytes();
+      if (in->getKind() == Kinded::Kind::ConstantKind) {
+        ret.constMemSize += size;
       } else {
-        // In this case, this input is neither a storage type node nor belongs
-        // to this subgraph. Therefore, when creating paritions, we need to add
-        // a PlaceHolder for the data from outside.
-        ret.inMemSize += nodeVal.getType()->getSizeInBytes();
+        // PlaceHolder for Input.
+        ret.inMemSize += size;
+      }
+      usedStorage.insert(in);
+      continue;
+    }
+
+    if (currNodes.count(N)) {
+      // This input is inside of the nodes set. Let node1 belongs to the nodes
+      // set and only node uses nodes1's output(i.e. node1 -> node, and if node1
+      // -> node2, node2 belongs to the current nodes set). Before node is
+      // added, the size of node1's ouput is added into outMemSize of this set's
+      // GraphMemInfo. But if node is added, node1's output size should be
+      // removed from outMemSize.
+      std::vector<Node *> users = getNodeUsers(N, nodeVal.getResNo());
```

Comment:
nit: If you change `getNodeUsers()` to take a NodeValue then you can just pass `nodeVal` here.